File: backend/app/models/git.py
# Git 操作相关的请求/响应模型：涵盖状态查询、暂存/取消暂存、提交、分支管理、
# 贮藏（stash）、丢弃变更、提交历史等 Git 面板功能所需的数据结构。
from pydantic import BaseModel


class GitFileChange(BaseModel):
    """单个文件的变更信息：路径、变更状态（如 M/A/D 等），以及可选的增删行数统计。"""

    path: str
    status: str
    insertions: int | None = None
    deletions: int | None = None


class GitStatusResponse(BaseModel):
    """Git 仓库状态：当前分支、领先/落后远程的提交数，以及已暂存/未暂存/未跟踪的文件变更列表。"""

    branch: str
    ahead: int
    behind: int
    staged: list[GitFileChange]
    unstaged: list[GitFileChange]
    untracked: list[GitFileChange]


class GitStageRequest(BaseModel):
    """暂存文件请求：指定项目及要暂存（git add）的文件路径列表。"""

    project_id: str
    paths: list[str]


class GitUnstageRequest(BaseModel):
    """取消暂存请求：指定项目及要取消暂存（git reset）的文件路径列表。"""

    project_id: str
    paths: list[str]


class GitCommitRequest(BaseModel):
    """提交请求：项目、提交信息，以及是否为修订上一次提交（amend）。"""

    project_id: str
    message: str
    amend: bool = False


class GitProjectRequest(BaseModel):
    """仅需项目标识的通用 Git 请求（如拉取/推送等无额外参数的操作）。"""

    project_id: str


class GitStashRequest(BaseModel):
    """贮藏操作请求：action 指定具体动作（如 push/pop 等），默认为贮藏当前变更。"""

    project_id: str
    action: str = "push"


class GitDiscardRequest(BaseModel):
    """丢弃变更请求：指定项目及要放弃修改（还原为最近提交状态）的文件路径列表。"""

    project_id: str
    paths: list[str]


class GitSimpleResponse(BaseModel):
    """通用的简单操作结果响应：是否成功，失败时附带错误信息。"""

    success: bool
    error: str | None = None


class GitBranchItem(BaseModel):
    """单个分支信息：名称、是否为当前所在分支、是否为远程分支。"""

    name: str
    is_current: bool
    is_remote: bool


class GitBranchListResponse(BaseModel):
    """分支列表响应：所有分支及当前所在分支名称。"""

    branches: list[GitBranchItem]
    current: str


class GitBranchCreateRequest(BaseModel):
    """创建分支请求：分支名称，checkout 指定创建后是否立即切换到该分支。"""

    project_id: str
    name: str
    checkout: bool = True


class GitBranchDeleteRequest(BaseModel):
    """删除分支请求：force 指定是否强制删除（忽略未合并的变更）。"""

    project_id: str
    name: str
    force: bool = False


class GitBranchSwitchRequest(BaseModel):
    """切换分支请求：切换到指定名称的分支。"""

    project_id: str
    name: str


class GitLogCommit(BaseModel):
    """单条提交记录：完整哈希、短哈希、作者、提交日期、提交信息。"""

    hash: str
    short_hash: str
    author: str
    date: str
    message: str


class GitLogResponse(BaseModel):
    """提交历史查询响应：提交记录列表。"""

    commits: list[GitLogCommit]


class GitLogRequest(BaseModel):
    """提交历史查询请求：max_count 限制返回的最大提交条数。"""

    project_id: str
    max_count: int = 50

File: frontend/src/types/git.ts
// 文件功能：Git 操作相关类型定义
// 文件描述：定义 git 状态码、文件变更、分支信息、提交记录等类型，以及对应的接口响应结构，
//          供前端 git 面板（状态/分支/提交历史展示）使用
// 核心逻辑：围绕 git status / branch / log 三类查询组织类型；staged/unstaged/untracked 对应
//          git 工作区三种文件变更分类
// git 变更状态码：M(修改) / A(新增) / D(删除) / U(未合并/冲突) / R(重命名)
export type GitStatusCode = 'M' | 'A' | 'D' | 'U' | 'R'

/**
 * 函数名：isValidGitStatusCode
 * 入参：
 *   - value (unknown): 待校验的值
 * 功能：类型收窄守卫，判断给定值是否为合法的 GitStatusCode
 * 运行逻辑：判断是否为字符串且属于 ['M', 'A', 'D', 'U', 'R'] 集合
 * 出参：boolean（类型谓词）- true 表示 value 可安全当作 GitStatusCode 使用
 */
export function isValidGitStatusCode(value: unknown): value is GitStatusCode {
  return typeof value === 'string' && ['M', 'A', 'D', 'U', 'R'].includes(value)
}

// 单个文件的 git 变更记录
export interface GitFileChange {
  path: string // 文件相对项目根目录的路径
  status: GitStatusCode
  insertions?: number | null // 新增行数，未统计时为 null/undefined
  deletions?: number | null // 删除行数，未统计时为 null/undefined
}

// 分支基本信息：名称及与其上游分支的领先/落后提交数
export interface GitBranchInfo {
  name: string
  ahead: number // 领先上游分支的提交数
  behind: number // 落后上游分支的提交数
}

// git status 接口响应：当前分支及三类文件变更（已暂存/未暂存/未跟踪）
export interface GitStatusResponse {
  branch: string
  ahead: number
  behind: number
  staged: GitFileChange[] // 已 git add 暂存的变更
  unstaged: GitFileChange[] // 已跟踪但未暂存的变更
  untracked: GitFileChange[] // 尚未被 git 跟踪的新文件
}

// 通用 git 操作结果响应（如 commit/push/pull 等一次性操作）
export interface GitSimpleResponse {
  success: boolean
  error: string | null
}

// 单个分支条目
export interface GitBranchItem {
  name: string
  is_current: boolean // 是否为当前所在分支
  is_remote: boolean // 是否为远程分支
}

// 分支列表接口响应
export interface GitBranchListResponse {
  branches: GitBranchItem[]
  current: string // 当前分支名
}

// 单条 git 提交记录
export interface GitLogCommit {
  hash: string // 完整提交哈希
  short_hash: string // 短哈希（用于展示）
  author: string
  date: string
  message: string // 提交信息
}

// 提交历史接口响应
export interface GitLogResponse {
  commits: GitLogCommit[]
}
